### PR TITLE
Exclude boundary nodes from test_step spatial loss maps

### LINKED ISSUE #569: [Bug] spatial_loss_maps in test_step computed without interior_mask, inconsistent with all other loss calls
**Description**

In `ar_model.py`, `test_step` computes `spatial_loss_maps` at L452-454 without passing `mask=self.interior_mask_bool`:

```python
spatial_loss = self.loss(
    prediction, target, pred_std, average_grid=False
)
```

Every other call to `self.loss` in the module passes `mask=self.interior_mask_bool`:

- `training_step` (L311)
- `validation_step` (L354)
- `test_step` main loss (L408)

The per-variable metrics (`mse`, `mae`) computed just above (L435-441) also pass the mask. This means the spatial loss maps include boundary grid nodes that are excluded everywhere else.

**Impact**

- The saved `mean_spatial_loss.pt` contains loss values at boundary nodes that the model never optimises for, which can mislead downstream error analysis.
- `plot_spatial_error` picks `vmin`/`vmax` from the full tensor, so boundary-node loss values can skew the colour scale.
- For LAM domains with boundary forcing the inconsistency is most pronounced; for global domains (no boundary) it has no practical effect.

**Minimal example**

```bash
python -m neural_lam.train_model --eval test --config_path <config> --load <ckpt>
# spatial_loss_maps saved to <logger_dir>/spatial_loss_maps/ include boundary values
```

**Proposed fix**

After computing spatial loss over the full grid (to preserve the `num_grid_nodes` shape needed by plotting), mask boundary nodes to NaN:

```python
spatial_loss = self.loss(
    prediction, target, pred_std, average_grid=False
)
spatial_loss[..., ~self.interior_mask_bool] = float("nan")
```

And use `torch.nanmean` instead of `torch.mean` when averaging the maps in `on_test_epoch_end`.

### PR DESCRIPTION
**Pull request by NoiceHax**
## Describe your changes

`test_step` built its spatial loss map over the whole grid, while every other
loss and metric call in `ForecasterModule` restricts itself to the interior via
`interior_mask_bool`. `ARForecaster` overwrites the boundary nodes with the true
boundary state, so their loss is always ~0, and those zeros were dragging down
the saved `mean_spatial_loss.pt` and flattening the colour range of the plotted
loss maps.

This NaN-masks the boundary right after the loss call and reduces with
`torch.nanmean` in `on_test_epoch_end`. Passing `mask=self.interior_mask_bool`
into `self.loss` is not an option: `metrics.mask_and_reduce_metric`
index-selects with `metric_entry_vals[..., mask, :]`, which shrinks the grid
dimension and breaks the shape `vis.plot_spatial_error` reshapes back onto the
grid. `vis.py` needs no change, it already uses `np.nanmin`/`np.nanmax` and
matplotlib draws NaN cells transparent.

The fix is @RajdeepKushwaha5's from #568, following @sadamov's review there.
#568 edits `neural_lam/models/ar_model.py`, which no longer exists after the
`ForecasterModule` split in #208, so it cannot be rebased onto main. The new
test is the first coverage of `test_step` / `on_test_epoch_end`. It fails on
main and passes here.

No new dependencies. Written with help from Claude Code.

## Issue Link

closes #569

## Type of change

- [x] ðŸ› Bug fix (non-breaking change that fixes an issue)
- [ ] âœ¨ New feature (non-breaking change that adds functionality)
- [ ] ðŸ’¥ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] ðŸ“– Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [x] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes - not needed, nothing user-facing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [x] the code is readable
- [x] the code is well tested
- [x] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [x] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [x] PR is up to date with the base branch
- [x] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [x] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -50,6 +50,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- Exclude boundary nodes from the spatial loss maps computed in `test_step`, so the plotted loss maps and the saved `mean_spatial_loss.pt` cover the interior only, consistent with every other loss and metric call in `ForecasterModule` [\#720](https://github.com/mllam/neural-lam/pull/720) @NoiceHax
```
Comment (sadamov) [inline CHANGELOG.md]: The fix is originally @RajdeepKushwaha5's from #568, could we credit you both here?

```suggestion
- Exclude boundary nodes from the spatial loss maps computed in `test_step`, so the plotted loss maps and the saved `mean_spatial_loss.pt` cover the interior only, consistent with every other loss and metric call in `ForecasterModule` [\#720](https://github.com/mllam/neural-lam/pull/720) @RajdeepKushwaha5 @NoiceHax
```

**Review by sadamov — APPROVED** [review]
Thanks for picking this up. I verified locally: the new test fails on main with the boundary nodes sitting at exactly `0.0`, and passes with the fix, so it is real regression coverage.

This stays relevant after #636, which only deprecates `crop_to_interior` and always crops the axes to the interior bbox. The `vrange=None` call sites still take `vmin`/`vmax` from the full grid, so the boundary zeros keep pinning `vmin` to 0, and `mean_spatial_loss.pt` is still saved over the whole grid.

One request on the CHANGELOG below, and then this is ready from my side.

File: CHANGELOG.md (hunk 1 above)
Comment (NoiceHax) [reply CHANGELOG.md]: Absolutely, Its only fair to give credit where its due.

File: neural_lam/models/module.py [hunk 2]
```diff
@@ -611,6 +611,11 @@ def test_step(self, batch, batch_idx):
         spatial_loss = self.loss(
             prediction, target_states, pred_std, average_grid=False
         )
+        # Exclude boundary nodes, consistent with the loss used in
+        # training/validation. NaN-mask in place so the tensor keeps its
+        # (B, pred_steps, num_grid_nodes) shape for downstream plotting,
+        # and is reduced with `torch.nanmean` in `on_test_epoch_end`.
```
Comment (sadamov) [inline neural_lam/models/module.py]: remove
```suggestion
```

File: tests/test_training.py [hunk 3]
```diff
@@ -227,3 +231,115 @@ def all_gather(self, tensor, sync_grads=False):
...
+        mask_2d = np.zeros((n_points_1d, n_points_1d), dtype=int)
+        mask_2d[0, :] = 1
+        mask_2d[-1, :] = 1
+        mask_2d[:, 0] = 1
+        mask_2d[:, -1] = 1
+        # `grid_index` is the stacked ("x", "y") dimension, so a C-order
+        # flatten of an (x, y) array lines up with it.
+        return xr.DataArray(mask_2d.reshape(-1), dims=["grid_index"])
+
+
```
Comment (sadamov) [inline tests/test_training.py]:
```suggestion

@pytest.mark.slow
```

File: tests/test_training.py [hunk 4]
```diff
@@ -227,3 +231,115 @@ def all_gather(self, tensor, sync_grads=False):
...
+    def boundary_mask(self) -> xr.DataArray:
+        n_points_1d = int(np.sqrt(self.num_grid_points))
+        mask_2d = np.zeros((n_points_1d, n_points_1d), dtype=int)
+        mask_2d[0, :] = 1
+        mask_2d[-1, :] = 1
+        mask_2d[:, 0] = 1
+        mask_2d[:, -1] = 1
+        # `grid_index` is the stacked ("x", "y") dimension, so a C-order
+        # flatten of an (x, y) array lines up with it.
+        return xr.DataArray(mask_2d.reshape(-1), dims=["grid_index"])
```
Comment (sadamov) [inline tests/test_training.py]: `test_plotting.py:365` already frames a boundary this way, by assigning datastore.ds["boundary_mask"] and popping the cached_property. Could we lift that into tests/dummy_datastore.py and drop the subclass? #636 rewrites that plotting test too, so a shared helper saves a third copy.

something like_
```
def set_framed_boundary(
    datastore: DummyDatastore, width: int = 1
) -> np.ndarray:
    """Replace the random boundary mask with a frame of `width` cells."""
    shape = (datastore.grid_shape_state.x, datastore.grid_shape_state.y)
    mask = np.ones(shape, dtype=int)
    mask[width:-width, width:-width] = 0
    # `grid_index` is the stacked ("x", "y") dimension, so a C-order flatten
    # of an (x, y) array lines up with it.
    datastore.ds["boundary_mask"] = xr.DataArray(
        mask.reshape(-1), dims=["grid_index"]
    )
    datastore.__dict__.pop("boundary_mask", None)  # bust the cached_property
    return mask
```
Then here, which also drops the now unused cached_property, numpy and xarray imports:


```
def test_test_step_excludes_boundary_from_spatial_loss(tmp_path):
    ...
    datastore = DummyDatastore(n_grid_points=400, n_timesteps=10)
    set_framed_boundary(datastore)
```

**Review by sadamov — COMMENTED** [review]
@NoiceHax before merging with main I carried out my default list of review points and found some additional nitpicks. could you please have a look

**Comment by NoiceHax** [comment]
Addressed all the nitpicks.

- Dropped the redundant comment in module.py.
- Lifted set_framed_boundary into tests/dummy_datastore.py and dropped the subclass plus its now-unused numpy / xarray / cached_property imports. The plotting test can reuse the helper too.
- Marked the new test @pytest.mark.slow.
- Also added @RajdeepKushwaha5 to the CHANGELOG entry from the earlier request.

black and isort pass on the touched files. I still can't run the suite locally, so the fix and test logic are unchanged from what you verified.

**Review by sadamov — APPROVED** [review]
great! merging this fix now, thanks @NoiceHax and @RajdeepKushwaha5
